hnn_core/parallel_backends.py

```
@@ -0,0 +1,265 @@
+
+        # the assignments below need to be made after any forking
+        for idx in range(n_trials):
+            dpls.append(data[idx][0])
+            spikedata = data[idx][1]
+            net.spiketimes.append(spikedata[0])
+            net.spikegids.append(spikedata[1])
+            net.gid_dict = spikedata[2]  # only have one gid_dict
+
+        return dpls
+
+
+class MPI_backend(object):
+    """The MPI_backend class.
+
+    Parameters
+    ----------
+    n_procs : int | None
+        The number of processors (cores) to start MPI job over
+    mpi_cmd : str
+        The name of the mpi launcher executable. Will use 'mpiexec'
+        (openmpi) by default.
+
+    Attributes
+    ----------
+
+    n_jobs : int
+        The number of jobs to start in parallel (NOT SUPPORTED)
+    n_cores : int
+        The number of cores used by the backend
+    mpi_cmd_str : str
+        The string of the mpi command with number of cores and options
+
+    """
+    def __init__(self, n_jobs=1, n_procs=None, mpi_cmd='mpiexec'):
+        import psutil
+        import multiprocessing
+        import os
+        import hnn_core
+
+        try:
+            import mpi4py
+            mpi4py.__version__  # for flake8 test
+        except ImportError:
+            warn('mpi4py not installed. will run on single processor')
+            self.n_procs = 1
+
+        self._type = 'mpi'
+        self.n_jobs = n_jobs
+
+        physical_cores = psutil.cpu_count(logical=False)
+        logical_cores = multiprocessing.cpu_count()
+
+        oversubscribe = False
+        # trying to run on more than available logical cores?
+        if n_procs is None:
+            self.n_cores = logical_cores
+        else:
+            self.n_cores = n_procs
+            if n_procs > logical_cores:
+                oversubscribe = True
+
+        # obey limits set by scheduler
+        try:
+            scheduler_cores = len(os.sched_getaffinity(0))
+        except AttributeError:
```

instead of a try-except clause, I'd suggest using `hasattr`. It's a better pattern in general because you don't suppress errors

---

Fixed.

https://github.com/jonescompneurolab/hnn-core/blob/1cc993d6289ceb30b317a86813f4190a8205fddf/hnn_core/parallel_backends.py#L157-L160